core/src/raw/adapters/kv/api.rs

```diff
@@ -25,10 +26,69 @@ use crate::Capability;
 use crate::Scheme;
 use crate::*;
 
+/// ScanIter is the async iterator returned by `Adapter::scan`.
+pub trait Scan: Send + Sync + Unpin {
+    /// Fetch the next key in the current key prefix
+    ///
+    /// `None` means no further key will be returned
+    fn next(&mut self) -> impl Future<Output = Option<Result<String>>> + MaybeSend;
+}
+
+/// A noop implementation of Scan
+impl Scan for () {
+    async fn next(&mut self) -> Option<Result<String>> {
+        None
+    }
+}
+
+/// A ScanIterator implementation for all trivial non-async iterators
```

Comment:
I feel that this is unnecessary. The service should provide their own scanner when needed. Ideally, service should perform real IO inside scanner instead of in the `scan` call.

For example, mysql service should build connection / sending query inside the first `next` call in scanner.

Reply:
Hmm I currently use it to quickly refactor the code i.e. lots of these services just return a Vec. Also I think some of these services don't provide a batched-fetching method on Scan so it may be useful.

Do you have any idea on this?

Reply:
>  just return a Vec

Yes, this is part of the dark history I don't want to inherit.

---

> Hmm I currently use it to quickly refactor the code.

I understand your motivation. Let's keep it but make it internally contained and not exposed to users; we can refactor this part later.

Reply:
Great!